source/common/network/io_socket_handle_impl.cc

```diff
@@ -84,8 +84,14 @@ Api::IoCallUint64Result IoSocketHandleImpl::readv(uint64_t max_length, Buffer::R
     num_bytes_to_read += slice_length;
   }
   ASSERT(num_bytes_to_read <= max_length);
-  return sysCallResultToIoCallResult(Api::OsSysCallsSingleton::get().readv(
+  auto result = sysCallResultToIoCallResult(Api::OsSysCallsSingleton::get().readv(
       fd_, iov.begin(), static_cast<int>(num_slices_to_read)));
+
+  // Some tests try to read without initializing the file_event.
+  if (result.wouldBlock() && file_event_) {
+    file_event_->registerEventIfEmulatedEdge(Event::FileReadyType::Read);
+  }
```

Comment:
Currently the tax is 2 if statements and a dummy function call. 

I see two approaches:
1. Wait for #13418 to get merged (it seems that this will be soon) and I will create a new io_handle extension just for that. Change the `makeSocket` function to get the appropriate socket for each case.
2. Add the following change that would help the compile optimize this code away in POSIX:
```c++
void FileEventImpl::unregisterEventIfEmulatedEdge(uint32_t event) {
  if constexpr ( PlatformDefaultTriggerType == FileTriggerType::EmulatedEdge) {
  ASSERT(event & (Event::FileReadyType::Write | Event::FileReadyType::Read)) == event);
  if (trigger_ == FileTriggerType::EmulatedEdge) {
    auto new_event_mask = enabled_events_ & ~event;
    updateEvents(new_event_mask);
  }
 }
}
```

Reply:
I just validated on [godbolt](https://godbolt.org/z/GWenzz) on a smaller example that clang would actually optimize the function call and the if statements away in O3.

Reply:
Option 2 seems like a good compromise.  I would suggest also having some ASSERTs to ensure that EmulatedEdge is not used for IoHandles when PlatformDefaultTriggerType != FileTriggerType::EmulatedEdge

Reply:
added the assert in the `file_event_impl`